File: benchmarks/csharp/MerkleTrees.cs
using Godot;

public partial class MerkleTrees : Benchmark
{
    // Based on https://github.com/hanabi1224/Programming-Language-Benchmarks/blob/main/bench/algorithm/merkletrees/1.cs
    class TreeNode
    {
        public long? value;
        public long? hash;
        public TreeNode left;
        public TreeNode right;

        public TreeNode(long? value, TreeNode left = null, TreeNode right = null)
        {
            this.value = value;
            this.left = left;
            this.right = right;
        }

        public static TreeNode Create(int d)
        {
            return d == 0 ? new TreeNode(1L, null, null)
                          : new TreeNode(null, Create(d - 1), Create(d - 1));
        }

        public bool Check()
        {
            if (hash != null)
            {
                if (value != null)
                {
                    return true;
                }
                else if (left != null && right != null)
                {
                    return left.Check() && right.Check();
                }
            }
            return false;
        }

        public long GetHash()
        {
            if (hash.Value is long v)
            {
                return v;
            }
            return default;
        }

        public void CalHash()
        {
            if (hash == null)
            {
                if (value.HasValue)
                {
                    hash = value;
                }
                else if (left != null && right != null)
                {
                    left.CalHash();
                    right.CalHash();
                    hash = left.GetHash() + right.GetHash();
                }
            }
        }
    }

    const int MinDepth = 4;
    public static void CalculateMerkleTrees(int input)
    {
        int maxDepth = Mathf.Max(MinDepth + 2, input);

        int stretchDepth = maxDepth + 1;
        TreeNode stretchTree = TreeNode.Create(stretchDepth);
        stretchTree.CalHash();
        GD.Print($"stretch tree of depth {stretchDepth}\t root hash: {stretchTree.GetHash()} check: {stretchTree.Check().ToString().ToLowerInvariant()}");

        TreeNode longLivedTree = TreeNode.Create(maxDepth);
        int maxPlusMinDepth = maxDepth + MinDepth;
        for (int depth = MinDepth; depth < maxDepth; depth += 2)
        {
            int iterations = 1 << (maxPlusMinDepth - depth);
            long sum = 0;
            for (int i = 0; i < iterations; i++)
            {
                TreeNode tree = TreeNode.Create(depth);
                tree.CalHash();
                sum += tree.GetHash();
            }
            GD.Print($"{iterations}\t trees of depth {depth}\t root hash sum: {sum}");
        }

        longLivedTree.CalHash();
        GD.Print($"long lived tree of depth {maxDepth}\t root hash: {longLivedTree.GetHash()} check: {longLivedTree.Check().ToString().ToLowerInvariant()}");
    }

    public void BenchmarkMerkleTrees13()
    {
        CalculateMerkleTrees(13);
    }

    public void BenchmarkMerkleTrees15()
    {
        CalculateMerkleTrees(15);
    }

    public void BenchmarkMerkleTrees18()
    {
        CalculateMerkleTrees(18);
    }
}

File: gdextension/src/benchmarks/merkle_trees.cpp
#include "merkle_trees.h"

#include <godot_cpp/core/class_db.hpp>
#include <godot_cpp/variant/utility_functions.hpp>

using namespace godot;

void CPPBenchmarkMerkleTrees::_bind_methods() {
	ClassDB::bind_method(D_METHOD("benchmark_merkle_trees_13"), &CPPBenchmarkMerkleTrees::benchmark_merkle_trees_13);
	ClassDB::bind_method(D_METHOD("benchmark_merkle_trees_15"), &CPPBenchmarkMerkleTrees::benchmark_merkle_trees_15);
	ClassDB::bind_method(D_METHOD("benchmark_merkle_trees_18"), &CPPBenchmarkMerkleTrees::benchmark_merkle_trees_18);
}

CPPBenchmarkMerkleTrees::TreeNode *CPPBenchmarkMerkleTrees::make_tree(int depth) {
	if (depth > 0) {
		depth -= 1;
		return new TreeNode(-1, make_tree(depth), make_tree(depth));
	}
	return new TreeNode(1, nullptr, nullptr);
}

void CPPBenchmarkMerkleTrees::calculate_merkle_trees(int input) {
	int max_depth = Math::max(min_depth + 2, input);

	int stretch_depth = max_depth + 1;
	TreeNode *stretch_tree = make_tree(stretch_depth);
	stretch_tree->cal_hash();
	UtilityFunctions::print("stretch tree of depth ", stretch_depth, "\t root hash: ", stretch_tree->hash, "\t check: ", stretch_tree->check());
	delete stretch_tree;

	TreeNode *long_lived_tree = make_tree(max_depth);
	int max_plus_min_depth = max_depth + min_depth;
	for (int depth = min_depth; depth < max_depth; depth += 2) {
		int iterations = 1 << (max_plus_min_depth - depth);
		int sum = 0;
		for (int i = 0; i < iterations; i++) {
			TreeNode *tree = make_tree(depth);
			tree->cal_hash();
			sum += tree->hash;
			delete tree;
		}

		UtilityFunctions::print(iterations, "\t trees of depth ", depth, "\t root hash sum: ", sum);
	}
	long_lived_tree->cal_hash();
	UtilityFunctions::print("long lived tree of depth ", input, "\t root hash: ", long_lived_tree->hash, "\t check: ", long_lived_tree->check());
	delete long_lived_tree;
}

void CPPBenchmarkMerkleTrees::benchmark_merkle_trees_13() {
	calculate_merkle_trees(13);
}

void CPPBenchmarkMerkleTrees::benchmark_merkle_trees_15() {
	calculate_merkle_trees(15);
}

void CPPBenchmarkMerkleTrees::benchmark_merkle_trees_18() {
	calculate_merkle_trees(18);
}

CPPBenchmarkMerkleTrees::CPPBenchmarkMerkleTrees() {}

CPPBenchmarkMerkleTrees::~CPPBenchmarkMerkleTrees() {}
